R/appMetadata.R

```diff
@@ -60,6 +61,50 @@ appMetadata <- function(appDir,
   )
 }
 
+checkAppLayout <- function(appDir, appPrimaryDoc = NULL) {
+  appFilesBase <- tolower(list.files(appDir))
+  wwwFiles <- tolower(list.files(file.path(appDir, "www/")))
+
+  primaryIsRScript <- identical(tolower(tools::file_ext(appPrimaryDoc)), "r")
+
+  # check for single-file app collision
+  if (primaryIsRScript && "app.r" %in% appFilesBase) {
+    cli::cli_abort(
+      "Project must not contain both {.file app.R} and a single-file Shiny app."
+    )
+  }
+
+  # Do some checks for a valid application structure
+  satisfiedLayouts <- c(
+    shinyAndUi = all(c("server.r", "ui.r") %in% appFilesBase),
+    shinyAndIndex = "server.r" %in% appFilesBase && "index.html" %in% wwwFiles,
+    app = primaryIsRScript || any("app.r" %in% appFilesBase),
+    Rmd = any(grepl(glob2rx("*.rmd"), appFilesBase)),
+    Qmd = any(grepl(glob2rx("*.qmd"), appFilesBase)),
+    static = any(grepl("(?:html?|pdf)$", appFilesBase)),
+    plumber = any(c("entrypoint.r", "plumber.r") %in% appFilesBase),
+    tensorflow = length(c(
+      Sys.glob(file.path(appDir, "*", "saved_model.pb*")),
+      Sys.glob(file.path(appDir, "saved_model.pb*"))
+    )) > 0
+  )
+
+  if (any(satisfiedLayouts)) {
+    return()
+  }
+
+  cli::cli_abort(c(
+    "Cancelling deployment: invalid project layout.",
+    i = "Expecting one of the following publication types:",
+    " " = "1. A Shiny app with `app.R` or `server.R` + `ui.R`",
+    " " = "2. An R Markdown (`.Rmd`) or Quarto (`.qmd`) document.",
```

Comment:
nit: Maybe pluralize "document".

Reply:
It looks like it only applies to shinyapps.io and ends up running the `deployApplication` against whatever we have stored as `bundle <- application$deployment$bundle`.

What this probably means is that rsconnect should run the "deployApplication" endpoint against whatever was the most recent bundle. I'm not sure about its history or if it still has a purpose.

@kippandrew, do you know if the `upload=FALSE` workflow (where the deploy happens without an upload) is valuable?

For Connect, the server supports such a thing, even though it's not made available today by rsconnect. If you call the "deploy" endpoint without a bundle identifier, Connect will "activate" the most recent bundle. That can be helpful if, for example, a systems change has been made that allows some package to successfully install. Typically, folks just upload again.